File: 23_merge_k_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        /*
         * Use Heap to solve. O(nlog(k))
         */
         if (lists == null || lists.length == 0) return null;
         
         Queue<ListNode> heap = new PriorityQueue(lists.length, new Comparator<ListNode>() {
            @Override
            public int compare(ListNode l1, ListNode l2) {
                return l1.val - l2.val;
            }}
         );
         for (ListNode n : lists) if (n != null) heap.add(n);
         
         ListNode dummy = new ListNode(0);
         ListNode tail = dummy;
         while (!heap.isEmpty()) {
             tail.next = heap.poll();
             tail = tail.next;
             if (tail.next != null) heap.add(tail.next);
         }
         return dummy.next;
    }
}

File: 23_merge_k_sorted_lists.py
"""
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity. 
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    """
    idea:
        divide and conquer
        Time: O(nlog(k))
    """
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if not lists:
            return
        if len(lists) == 1:
            return lists[0]

        mid = len(lists) / 2
        left = self.mergeKLists(lists[:mid])
        right = self.mergeKLists(lists[mid:])
        return self.merge(left, right)

    def merge(self, left, right):
        head = None
        curr = None
        while left != None or right != None:
            if left == None:
                if head == None:
                    head = right
                    curr = right
                    right = right.next
                else:
                    curr.next = right
                    curr = right
                    right = right.next
                    curr.next = None
            elif right == None:
                if head == None:
                    head = left
                    curr = left
                    left = left.next
                else:
                    curr.next = left
                    curr = left
                    left = left.next
                    curr.next = None
            elif left.val > right.val:
                if head == None:
                    head = right
                    curr = right
                    right = right.next
                else:
                    curr.next = right
                    curr = right
                    right = right.next
                    curr.next = None
            else:
                if head == None:
                    head = left
                    curr = left
                    left = left.next
                else:
                    curr.next = left
                    curr = left
                    left = left.next
                    curr.next = None
        return head


